### PR TITLE
Detach hiddens and add test

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

This was accidentally removed with https://github.com/PyTorchLightning/pytorch-lightning/pull/8145

Updated a test to avoid regressions

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

**Review by ethanwharris — APPROVED** [review]

File: tests/trainer/logging_/test_train_loop_logging.py [hunk 1]
```diff
@@ -261,7 +261,9 @@ def __init__(self):
 
         def training_step(self, batch, batch_idx, hiddens):
             assert hiddens == self.test_hidden, "Hidden state not persistent between tbptt steps"
-            self.test_hidden = torch.rand(1)
+            if hiddens is not None:
+                assert hiddens.grad_fn is None
+            self.test_hidden = torch.tensor(2., requires_grad=True).pow(2)
```
Comment (awaelchli) [inline tests/trainer/logging_/test_train_loop_logging.py:266]: great! verified it fails without the fix. 
either it was there forever or I had broken it with loop refactor :(((

**Review by awaelchli — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]
